pkg/metrics/providers/keptn.go

```diff
@@ -0,0 +1,239 @@
+	}
+	return &KeptnProvider{
+		client:          client,
+		analysisTimeout: 10 * time.Second,
+	}, nil
+}
+
+// RunQuery fetches the value of a KeptnMetric or Analysis,
+// based on the selector provided in the query.
+// The format of the selector is the following:
+// <keptnmetric|analysis>/<namespace>/<resourceName>/<duration>/<arguments>
+func (k *KeptnProvider) RunQuery(query string) (float64, error) {
+	queryObj, err := parseQuery(query)
+	if err != nil {
+		return 0, err
+	}
+
+	switch queryObj.GroupVersionResource.Resource {
+	case keptnMetricsResourceName:
+		return k.queryKeptnMetric(queryObj)
+	case analysisResourceName:
+		return k.queryKeptnAnalysis(queryObj)
+	default:
+		return 0, errors.New("unsupported query")
+	}
+
+}
+
+func (k *KeptnProvider) IsOnline() (bool, error) {
+	// TODO should we check for the keptn deployment to be up and running in the cluster?
+	return true, nil
+}
+
+func (k *KeptnProvider) queryKeptnMetric(queryObj *queryObject) (float64, error) {
+	get, err := k.client.Resource(queryObj.GroupVersionResource).
+		Namespace(queryObj.Namespace).
+		Get(
+			context.Background(),
+			queryObj.ResourceName,
+			v1.GetOptions{},
+		)
+
+	if err != nil {
+		return 0, fmt.Errorf("could not retrieve KeptnMetric %s/%s: %w", queryObj.Namespace, queryObj.ResourceName, err)
+	}
+
+	if status, ok := get.Object["status"]; ok {
+		if statusObj, ok := status.(map[string]interface{}); ok {
+			if value, ok := statusObj["value"].(string); ok {
+				floatValue, err := strconv.ParseFloat(value, 64)
+				if err != nil {
+					return 0, fmt.Errorf("could not parse value of KeptnMetric %s/%s to float: %w", queryObj.Namespace, queryObj.ResourceName, err)
+				}
+				return floatValue, nil
+			}
+		}
+	}
+	return 0, fmt.Errorf("could not retrieve KeptnMetric - no value found int resource %s/%s", queryObj.Namespace, queryObj.ResourceName)
+}
+
+func (k *KeptnProvider) queryKeptnAnalysis(obj *queryObject) (float64, error) {
+	analysis := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": fmt.Sprintf("metrics.keptn.sh/%s", apiVersion),
+			"kind":       "Analysis",
+			"metadata": map[string]interface{}{
+				"name":      fmt.Sprintf("%s-%s", obj.ResourceName, uuid.New().String()[:6]),
+				"namespace": obj.Namespace,
+			},
+			"spec": map[string]interface{}{
+				"analysisDefinition": map[string]interface{}{
+					"name": obj.ResourceName,
+				},
+				"timeframe": map[string]interface{}{
+					"recent": obj.DurationString,
+				},
+				"args": obj.Arguments,
+			},
+		},
+	}
+
+	// set the timeout to 10s - this will give Keptn enough time to reconcile the Analysis
+	// and store the result in the status of the resource created here.
+	ctx, cancel := context.WithTimeout(context.Background(), k.analysisTimeout)
+	defer cancel()
+
+	createdAnalysis, err := k.client.
+		Resource(obj.GroupVersionResource).
+		Namespace(obj.Namespace).
+		Create(ctx, analysis, v1.CreateOptions{})
```

Comment:
I think we should delete the generated objects when we are done reading the result so we don't fill etcd.